### PR TITLE
CA-395789: Add polling to cluster health state update

### PR DESCRIPTION
**Pull request by Vincent-lau**
It was observed that the corosync-notifyd might send a notification when the data shown in corosync-quorumtool has not been updated. This causes xapi-clusterd to return the out-of-date cluster information to xapi. Xapi will not have a chance of updating these states again until there is a further change in the cluster (which may not happen for a long time).

Now add polling in xapi to update the cluster info every 5 minutes, making sure that in the worst case, we still get up-to-date information after a 5-minute delay.

### CODE REVIEW

File: ocaml/xapi/xapi_clustering.ml [hunk 1]
```diff
@@ -535,10 +535,10 @@ module Watcher = struct
 
   let cluster_change_watcher : bool Atomic.t = Atomic.make false
 
-  (* this is the time it takes for the update request to time out. It is ok to set
+  (* This is the time it takes for the update request to time out. It is ok to set
      it to a relatively long value since the call will return immediately if there
-     is an update *)
-  let cluster_change_interval = Mtime.Span.min
+     is an update. Here we wait for 5 minutes *)
+  let cluster_change_interval = Clock.Timer.s_to_span 300. |> Option.get
```
Comment (psafont) [inline ocaml/xapi/xapi_clustering.ml:541]:
```suggestion
 let cluster_change_interval = Mtime.Span.(5 * min)
```

**Review by psafont — APPROVED** [review]

File: ocaml/xapi/xapi_clustering.ml [hunk 2]
```diff
@@ -535,10 +535,10 @@ module Watcher = struct
 
   let cluster_change_watcher : bool Atomic.t = Atomic.make false
 
-  (* this is the time it takes for the update request to time out. It is ok to set
+  (* This is the time it takes for the update request to time out. It is ok to set
      it to a relatively long value since the call will return immediately if there
-     is an update *)
-  let cluster_change_interval = Mtime.Span.min
+     is an update. *)
+  let cluster_change_interval = Mtime.Span.(5 * min)
```
Comment (gangj) [inline ocaml/xapi/xapi_clustering.ml:541]: May I know how the 5 minutes is determined for the updates check timeout, thank you!

File: ocaml/xapi/xapi_clustering.ml [hunk 3]
```diff
@@ -562,9 +562,18 @@ module Watcher = struct
...
-          ()
+      | Error (InternalError "UPDATES.Timeout") -> (
+        (* UPDATES.get timed out, this is normal.  *)
+        match find_cluster_host ~__context ~host with
+        | Some ch ->
+            let cluster = Db.Cluster_host.get_cluster ~__context ~self:ch in
+            (* CA-395789: We send a query to xapi-clusterd to fetch the latest state
+               anyway in case there is a race and the previous update did not give the
+               most up-to-date information *)
+            on_corosync_update ~__context ~cluster ["routine updates"]
```
Comment (gangj) [inline ocaml/xapi/xapi_clustering.ml:573]: The log in `on_corosync_update` will print:
`Received 1 updates from corosync_notifyd`
But in this case, there is no updates found, how about call `on_corosync_update` with an empty updates list?

File: ocaml/xapi/xapi_clustering.ml [hunk 4]
```diff
@@ -562,9 +562,18 @@ module Watcher = struct
...
+        | None ->
+            ()
```
Comment (gangj) [inline ocaml/xapi/xapi_clustering.ml:575]: Seems the logic here is the same as above case, how about add function for this? Then we can add comments before calling the function.

**Review by gangj — APPROVED** [review]

File: ocaml/xapi/xapi_clustering.ml (hunk 3 above)
Comment (Vincent-lau) [reply ocaml/xapi/xapi_clustering.ml:573]: I have adjusted the debug message according to the arguments

File: ocaml/xapi/xapi_clustering.ml (hunk 2 above)
Comment (gangj) [reply ocaml/xapi/xapi_clustering.ml:544]: No, thank you! Just for curiosity~

**Review by gangj — APPROVED** [review]

File: ocaml/xapi/xapi_clustering.ml [hunk 5]
```diff
@@ -428,10 +428,13 @@ let compute_corosync_max_host_failures ~__context =
 
 module Watcher = struct
   let on_corosync_update ~__context ~cluster updates =
-    debug
-      "%s: Received %d updates from corosync_notifyd, run diagnostics to get \
-       new state"
-      __FUNCTION__ (List.length updates) ;
+    if updates = ["routine updates"] then
```
Comment (gangj) [inline ocaml/xapi/xapi_clustering.ml:431]: Using a variable named like `routine_update` would be better.

**Review by lindig — APPROVED** [review]

**Pull request merged**